### ISSUE TITLE
Could this be made to read from a SD card?

### ISSUE THREAD

**Issue by Shell47** [samphonic/MoppyClassic#30]
Sorry to bother, but can this code be made independent of the PC?
So one can play the MIDI from and SD card connected to arduino?
Thank you!

**Comment by samphonic** [comment]
Early on in my experiments for this project, I tried to have the Arduino accept MIDI messages over serial so that you could theoretically hook up any MIDI device (PC, keyboard, etc) to the Arduino and it would play the notes on the floppy drives.  However, what I found was that the effort required to read/decode the MIDI messages made the Arduino unable to keep up with stepping the drives accurately, and they stopped producing notes.  I've since learned some optimization techniques (using timer interrupts for example), but I'm still a bit skeptical that the Arduino could handle decoding MIDI messages, much less reading a MIDI file _and_ decoding the messages.

That said, there's definitely some options for PC-independent usage.  I think, in theory, if you _could_ get an Arduino to read MIDI files, you could have one Arduino decoding the file, and the other driving the floppy drives.  Or for that matter, a RaspberryPi running the Java software talking to an Arduino. 

For a single Arduino solution, you may be better off using a custom-formatted file that contains just the bare-minimum of information for the floppy drive operation (drive number, note period, on/off).  You could have the file be essentially a two-dimensional array where each index of the primary array is one eighth-note period of the song, and each secondary array is a list of note on/off events that happen during that eighth-note period.  

Alternatively, you could have the file be a constant stream of messages, where some messages ask the Arduino to wait a certain number of milliseconds before continuing.  This would probably save space, and might simplify the timing a bit as you don't have to read an array index every-x-milliseconds.

tl;dr: Probably not a MIDI file directly from an SD card, but nothing's impossible and there's definitely PC-free possibilities.
